File: solutions/java_solutions/p947.java
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;

public class p947 {
    // Most Stones Removed with Same Row or Column - Medium

    // On a 2D plane, we place stones at some integer coordinate points. Each
    // coordinate point may have at most one stone.

    // Now, a move consists of removing a stone that shares a column or row with
    // another stone on the grid.

    // What is the largest possible number of moves we can make?

    class Solution {
        public int removeStones(int[][] stones) {
            if (stones.length <= 1)
                return 0;
            Graph g = generateGraph(stones);
            return countRemovableStones(g);
        }

        Graph generateGraph(int[][] stones) {
            HashMap<Integer, ArrayList<Integer>> x = new HashMap<>();
            HashMap<Integer, ArrayList<Integer>> y = new HashMap<>();

            for (int i = 0; i < stones.length; i++) {
                addStoneToMap(x, stones[i][0], i);
                addStoneToMap(y, stones[i][1], i);
            }

            Graph g = new Graph(stones.length);
            addEdges(g, x);
            addEdges(g, y);
            return g;
        }

        void addStoneToMap(HashMap<Integer, ArrayList<Integer>> map, int k, int v) {
            ArrayList<Integer> stones;
            if (map.containsKey(k))
                stones = map.get(k);
            else {
                stones = new ArrayList<Integer>();
                map.put(k, stones);
            }
            stones.add(v);
        }

        void addEdges(Graph g, HashMap<Integer, ArrayList<Integer>> map) {
            for (int key : map.keySet()) {
                ArrayList<Integer> set = map.get(key);
                // not sure if I need to exhaustively add all edges, or if I can get away with
                // being lazy here
                for (int i = 0; i < set.size() - 1; i++) {
                    for (int j = i + 1; j < set.size(); j++) {
                        g.addEdge(set.get(i), set.get(j));
                    }
                }
            }
        }

        int countRemovableStones(Graph g) {
            int stones = 0;
            boolean[] visited = new boolean[g.s];
            for (int i = 0; i < g.s; i++)
                if (!visited[i]) {
                    stones++;
                    markBFS(g, i, visited);
                }
            return stones;
        }

        void markBFS(Graph g, int s, boolean[] visited) {
            Queue<Integer> q = new LinkedList<>();
            q.add(s);

            while (!q.isEmpty()) {
                int u = q.poll();
                visited[u] = true;
                for (int v : g.adj.get(u))
                    if (!visited[v])
                        q.add(v);
            }
        }

        class Graph {
            ArrayList<ArrayList<Integer>> adj;
            int s;

            Graph(int s) {
                this.s = s;

                adj = new ArrayList<>();
                for (int i = 0; i < s; i++) {
                    adj.add(new ArrayList<>());
                }
            }

            void addEdge(int u, int v) {
                adj.get(u).add(v);
                adj.get(v).add(u);
            }
        }
    }
}

File: solutions/cpp_solutions/p947.cpp
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <vector>
using namespace std;

// Most stones removed with same row or column - Medium

// On a 2D plane, we place n stones at some integer coordinate points. 
//  Each coordinate point may have at most one stone.

// A stone can be removed if it shares either the same row or the same column as another stone that has not been removed.

// Given an array stones of length n where stones[i] = [xi, yi] represents the location of the ith stone, 
//  return the largest possible number of stones that can be removed.

class Solution {
public:
    typedef unordered_map<int, unordered_set<int>> coordSet;
    
    // for every row, collect each stone in a list (value will be column)
    coordSet rows;
    // vice versa for columns
    coordSet cols;
        
    int removeStones(vector<vector<int>>& stones) {
        if (stones.size() <= 1) return 0;
        
        for (auto &pair : stones) {
            rows[pair[0]].insert(pair[1]);
            cols[pair[1]].insert(pair[0]);
        }
        
        int totalRemoved = 0;
        while (!rows.empty()){
            auto pair = rows.begin();
            int x = pair->first;
            int y = *(pair->second.begin());
            int currentCount = dfsCount(x, y);
            totalRemoved += currentCount;
        }
        
        return totalRemoved;
    }
    
    int dfsCount(int x, int y)
    {
        int count = -1;
        stack<pair<int, int>> toVisit;
        removeFromMaps(x, y);
        toVisit.push({x, y});
        
        while (!toVisit.empty()){
            ++count;
            pair<int, int> next = toVisit.top();
            toVisit.pop();
            
            // find the next x neighbor
            auto rowiter = rows.find(next.first);
            if (rowiter != rows.end())
            {
                int otherY = *(rowiter->second.begin());
                removeFromMaps(next.first, otherY);
                toVisit.push({next.first, otherY});
            }
            
            // find the next y neighbor
            auto coliter = cols.find(next.second);
            if (coliter != cols.end()){
                int otherX = *(coliter->second.begin());
                removeFromMaps(otherX, next.second);
                toVisit.push({otherX, next.second});
            }
        }
        
        return count;
    }
    
    void removeFromMaps(int x, int y)
    {
        rows[x].erase(y);
        if (rows[x].empty()) rows.erase(x);
        
        cols[y].erase(x);
        if (cols[y].empty()) cols.erase(y);
    }
};